Adding date_start and date_end to Course model and making other backend adjustments

ssciolla opened the pull request:
This PR makes modifications to `models.py`, `views.py`, and `admin.py` and creates a new migration, in an effort to resolve [Issue 343](https://github.com/tl-its-umich-edu/my-learning-analytics/issues/343). The changes are summarized below:

1) Two new fields -- `date_start` and `date_end` -- were added to the `Course` model in order to allow admins to manually enter a date range for a course. A helper method `get_course_date_range` was also added to the model so that if course dates are not provided, the dates associated with the related`AcademicTerm` are accessed. Additionally, the `term_id` field was changed to `term` to reflect that the field holds the `AcademicTerms` object instance, not an integer or string. The `id` field was also made `editable`, possibly eliminating some unwanted behavior; see [Issue 613](https://github.com/tl-its-umich-edu/my-learning-analytics/issues/613).

2) The `objects` field of the `AcademicTerms` model was removed, as were the `Manager` and `QuerySet` helper classes for the model, as their functionality had been replaced with the changes to `Course` and they were no longer being used. A helper method `get_correct_date_end` was also added to correct far-out end years (usually ten years into the future) to the same year as the start date.

3) A new migration file, `0010_auto_20190812_0924.py`, was added to reflect the changes outlined in 1. and 2. (The original PR included another migration file, `0010_auto_20190809_1625.py`.)

4) Changes were made to `views.py` to rely on the `get_course_date_range` method for determining the course's beginning and end dates. These modifications mean that course dates provided in the admin interface will affect the number of weeks that show up in the front end, for instance, through the Resources Accessed view.

5) Some additional fields were provided in the `list_display` field of the `CourseAdmin` class in `admin.py` to make the list appearance of courses in the admin interface more informative.

zqian commented on dashboard/migrations/0010_auto_20190809_1625.py:
```diff
@@ -0,0 +1,45 @@
+        migrations.AlterField(
+            model_name='course',
+            name='id',
+            field=models.BigIntegerField(db_column='id', primary_key=True, serialize=False, verbose_name='Unizin Course Id'),
```
Maybe drop the notion of "Unizin" here since MyLA might be used by non-Unizin schools?

ssciolla replied:
Done! See latest commit.

zqian commented on dashboard/models.py:
```diff
@@ -14,36 +14,28 @@
+    # # This should be replaced in the future via an API call so the terms have correct end years, or Canvas data adjusted
+    def get_correct_date_end(self):
+        if (self.date_end.year - self.date_start.year) > 1:
+            logger.debug(f'{self.date_end.year} - {self.date_start.year} greater than 1 so setting end year to match start year.')
```
Since the date_end field is changed, the logging level will need to be "info" at least, instead of "debug"

ssciolla replied:
Thanks for this correction! I'm still getting comfortable with the `logging` module. See latest commit.

zqian commented on dashboard/models.py:
```diff
@@ -159,18 +153,31 @@ def get_supported_courses(self):
 
 class Course(models.Model):
-    id = models.BigIntegerField(primary_key=True, verbose_name="Unizin Course Id", db_column='id', editable=False)
+    id = models.BigIntegerField(primary_key=True, verbose_name="Unizin Course Id", db_column='id')
```
Drop the notion of "Unizin"?

ssciolla replied:
Done! See latest commit.

zqian commented on dashboard/models.py:
```diff
@@ -159,18 +153,31 @@ def get_supported_courses(self):
+    id = models.BigIntegerField(primary_key=True, verbose_name="Unizin Course Id", db_column='id')
     canvas_id = models.BigIntegerField(verbose_name="Canvas Course Id", db_column='canvas_id')
-    term_id = models.ForeignKey(AcademicTerms, verbose_name="Term Id", on_delete=models.SET_NULL, db_column='term_id', null=True, db_constraint=False)
+    term = models.ForeignKey(AcademicTerms, verbose_name="Term Id", on_delete=models.SET_NULL, db_column="term_id", null=True, db_constraint=False)
```
Can the "verbose_name" field also be changed to "Term", instead of "Term Id"?

ssciolla replied:
Sure! See latest commit.

zqian commented on dashboard/views.py:
```diff
@@ -593,11 +587,10 @@ def is_weight_considered(course_id):
-    return df['date_start'].iloc[0]
+def get_course_start_date(course_id):
+    logger.info(get_course_start_date.__name__)
+    course_date_start = Course.objects.all().get(id=course_id).get_course_date_range()[0]
```
Does "Course.objects.all()" part means to retrieve all course objects from database, instead of just a "get" with course_id?

ssciolla replied:
From a performance perspective, I think Django would treat this query as the same as `Course.objects.get(id=course_id)`; see [Querysets are lazy](https://docs.djangoproject.com/en/2.2/topics/db/queries/#querysets-are-lazy). Either way though, I think what you suggested is more concise. I made the change in the latest commit.

zqian replied:
Good to know!

ssciolla commented:
FYI, in addition to the changes @zqian requested, I simplified the verbose names for dates on the `AcademicTerms` model, and I modified the name of a method in `views.py` -- `get_course_date_start` -- to be in line with the name of the `Course` model field.

jonespm commented on dashboard/models.py:
```diff
@@ -159,18 +153,31 @@ def get_supported_courses(self):
+            end = self.date_end
+        else:
+            end = self.term.get_correct_date_end()
+        return start, end
```
A namedtuple would make this easier to read and work with. It's an 'anti-pattern' to use an ambiguous index.

https://docs.quantifiedcode.com/python-anti-patterns/readability/not_using_named_tuples_when_returning_more_than_one_value.html

ssciolla replied:
That makes sense to me. See latest commit.

jonespm commented on dashboard/views.py:
```diff
@@ -593,11 +587,10 @@ def is_weight_considered(course_id):
-    return df['date_start'].iloc[0]
+def get_course_date_start(course_id):
+    logger.info(get_course_date_start.__name__)
+    course_date_start = Course.objects.get(id=course_id).get_course_date_range()[0]
```
Using namedtuple (above) would make this get_course_date_range().date_start

ssciolla replied:
See latest commit.

jonespm commented:
@ssciolla  Looks good, this was a really good Pycon session where I first saw namedtuple and how cool it was if you ever have time to watch! https://www.youtube.com/watch?v=wf-BqAjZb8M

zqian commented:
> @ssciolla Looks good, this was a really good Pycon session where I first saw namedtuple and how cool it was if you ever have time to watch! https://www.youtube.com/watch?v=wf-BqAjZb8M

Maybe dev should have a list of "favorite videos to watch when you got time" somewhere...
